core/session.ts

```diff
@@ -217,50 +216,66 @@ export class Session {
     });
   }
 
-  public resolve(name: string): string {
-    const table = this.tables[name];
-    const operation =
-      !!this.operations[name] && this.operations[name].hasOutput && this.operations[name];
+  public resolve(ref: Resolvable): string {
+    const allResolved = this.findActions(ref);
+    if (allResolved.length > 1) {
+      const msg =
+        "Ambiguous Action name: " +
+        ref +
+        ". Did you mean one of: [" +
+        allResolved.join(", ") +
+        "].";
+      this.compileError(new Error(msg));
+    }
+    const resolved = allResolved.length > 0 ? allResolved[0] : undefined;
 
-    if (table && table.proto.type === "inline") {
+    if (resolved && resolved instanceof table.Table && resolved.proto.type === "inline") {
       // TODO: Pretty sure this is broken as the proto.query value may not
       // be set yet as it happens during compilation. We should evalute the query here.
-      return `(${table.proto.query})`;
+      return `(${resolved.proto.query})`;
+    }
+    if (resolved && resolved instanceof Operation && !resolved.proto.hasOutput) {
+      this.compileError(
+        new Error("Actions cannot resolve operations which do not produce output.")
+      );
     }
 
-    const dataset = table || operation;
     // TODO: We fall back to using the plain 'name' here for backwards compatibility with projects that use .sql files.
     // In these projects, this session may not know about all actions (yet), and thus we need to fall back to assuming
     // that the target *will* exist in the future. Once we break backwards compatibility with .sql files, we should remove
     // the code that calls 'this.target(...)' below, and append a compile error if we can't find a dataset whose name is 'name'.
-    const target = dataset ? dataset.proto.target : this.target(name);
+
+    const target = resolved
+      ? resolved.proto.target
+      : this.target(typeof ref === "string" ? ref : ref.name);
     return this.adapter().resolveTarget(target);
   }
 
   public operate(name: string, queries?: OContextable<string | string[]>): Operation {
-    this.checkActionNameIsUnused(name);
     const operation = new Operation();
     operation.session = this;
-    operation.proto.name = name;
     operation.proto.target = this.target(name);
```

Comment:
these 4 lines are common to all actions. suggest:

```
private setNameAndTarget(action: IActionProto, name: string) {
  action.target = this.target(name);
  this.checkTargetIsUnused(action.target);
  action.name = `${action.target.schema}.${action.target.name}`;
}
```

(note that `IActionProto` will probably have to change to include the `target` field)

Reply:
Done